### ISSUE TITLE
Templating change in upcoming sets: "CARDNAME" -> "this <object>"

### ISSUE THREAD

**Issue by xenohedron** [magefree/mage#12791]
Compare [[Basri's Acolyte]] (current templating) to [[Felidar Savior]] (new templating in Foundations).

Not sure how we want to handle this, but it will be a massive source of text discrepancies.

**Comment by JayDi85** [comment]
You can write special check in verify tests for it like:
* if ref’s rules contain name <> card’s rules contain name then fail

This will catch all needed changes (cards lists) after oracle update. Maybe it will require some ignore (for words from rules) and short (from planeswalkers) worlds lists.

**Comment by JayDi85** [comment]
Looks like wizards still miss oracle update for old sets like. As example: Basri's Acolyte still use "When Basri's Acolyte enters". Same true for all old sets.

So auto-usage of "this creature" in ETB (see code example in 0d0661cc924a4423423914d100111450115e43cb) will gives more problems.

Wrong rules stats for ALL sets (+2090):
* before ETB fix: 1943
* after ETB fix: 4033

Wrong rules stats for DFT;FDN;DSK sets (-37):
* before ETB fix: 344
* after ETB fix: 307

**Comment by JayDi85** [comment]
Another bad example: [[Aatchik, Emerald Radian]] -- it's new card but use short name instead "this creature" logic

**Comment by JayDi85** [comment]
And another new card with this instead short name replace: [[Mu Yanling, Wind Rider]]

**Comment by JayDi85** [comment]
[[Shefet Archfiend]] from DFT don't use that replace. So there are not any single logic for it.

**Conclusions**

Rules auto-replace improved by 19269b22b0647d64767a843d896baf7ddf625a55. Now it gives -80 problems for `DFT;FDN`:
- before fix: 333 problems
- after fix: 253 problems

Added `EntersBattlefieldTriggeredAbility.ENABLE_TRIGGER_PHRASE_AUTO_FIX` (disabled by default).

All that code can be used to support global "{this}" replace someday. In the future that logic must be combined with `TriggeredAbilityImpl.replaceRuleText` -- the last one used to replace "{this}" on "it".

There are full mess of different cards. So it's impossible to create global logic for "{this}" replace and require per card/ability settings anyway.

**Comment by JayDi85** [comment]
Scryfall has a good and detailed article about "this" and "cardname" usages. Briefly: it depends on card's legendary and ability's zone usages in most use cases with many exceptions, ~12k cards must be changed
https://scryfall.com/blog/errata-notice-aetherdrift-231

**Comment by xenohedron** [comment]
Yeah I agree - I started looking into it but I think it's going to require a rework of how "{this}" is handled. I think we should leave all existing text generation alone using "{this}" and instead change the method that replaces it - currently just with the cardname at the renderer level. It won't be that hard to write logic to handle the vast majority of cases, but it'll need to check the card characteristics. Only a few effects will need text generation adjusted - CDAs need to use a new "{cardname}" instead of "{this}", and type changing effects need to use "this permanent" instead of "{this}". Otherwise it should all be handled at the Ability level generating a rule, not the Effect level. Probably when an Ability is added to a card, it needs to get the card description at that time and store it in the Ability, rather than trying to pass a Card object to effects for text gen

The verify test as it currently stands isn't compatible with this approach, because it only has the cardname to substitute.

So suggested order of operations from my perspective:
- Revert 19269b22 and 0d0661cc (not the right approach)
- Change ability rule generation to substitute cardname for {this}, instead of leaving that to the renderer. (This solves the issue with the verify check.)
- Change the substitution to use the card characteristics logic instead of the bare cardname
- Add a second substitution that always uses the card name and update CDAs effects to use it instead
- Update type changing effect texts
- Check verify output, evaluate discrepancies, and think about best approach to solve them. (If no fundamental problems, can merge and then clean up the rest later, but if fundamental problems, may need to abandon this direction and try a different way.)

Thoughts?

**Comment by xenohedron** [comment]
Draft of code to generate description (based on scryfall blog post, but haven't been able to test output yet)

```java

private static final List<SubType> selfRefNamedSubtypes = Arrays.asList(
            SubType.EQUIPMENT,
            SubType.VEHICLE,
            SubType.AURA,
            SubType.CLASS,
            SubType.SAGA,
            SubType.SIEGE
    );

    /**
     * Returns the description of the card for rules text that references itself.
     * Applies to abilities that function only on the battlefield.
     * Does not account for every exception.
     * <a href="https://scryfall.com/blog/errata-notice-aetherdrift-231">Details here</a>
     */
    @Override
    public String getDescription() {
        if (!isPermanent()) {
            return getName();
        }
        if (isPlaneswalker()) {
            // try to ref by planeswalker name which is subtype
            return getSubtype().stream()
                    .filter(o -> SubType.getPlaneswalkerTypes().contains(o))
                    .findFirst()
                    .map(SubType::getDescription)
                    .orElse(getName());
        }
        if (isLegendary()) {
            // Generate shortname for legendary permanent (other than planeswalker)
            List<String> parts = Arrays.asList(getName().split("(, | of | the | of the )"));
            if (parts.size() > 1) {
                return parts.get(0);
            } else {
                return getName();
            }
        }
        if (isCreature()) {
            return "this creature";
        }
        if (isLand()) {
            return "this land";
        }
        for (SubType subType : selfRefNamedSubtypes) {
            if (getSubtype().contains(subType)) {
                return "this " + subType.getDescription();
            }
        }
        if (isBattle()) {
            return "this battle";
        }
        if (isEnchantment()) {
            return "this enchantment";
        }
        if (isArtifact()) {
            return "this artifact";
        }
        return "this permanent";
    }

```

**Comment by theelk801** [comment]
here's something I wrote a while back before stashing it and not coming back to it:

<img width="565" alt="Image" src="https://github.com/user-attachments/assets/8b52a684-0f5b-4443-80a8-b63b8cfd1068" />

an annoying thing about this is that because "this" isn't capitalized you have to update the capitalization is some places, which I did like this:

<img width="541" alt="Image" src="https://github.com/user-attachments/assets/60b384f3-f2cd-4a4d-8f29-82d21961c486" />

This was only in the verify text test but you get the idea
